jax/_src/prng.py

```diff
@@ -0,0 +1,475 @@
+from jax.interpreters import xla
+from jax._src.pprint_util import pp, vcat
+from jax._src.util import prod
+
+
+UINT_DTYPES = {
+    8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}  # type: ignore[has-type]
+
+# -- PRNG implementation interface --
+
+class PRNGImpl(NamedTuple):
+  key_shape: core.Shape
+  seed: Callable
+  split: Callable
+  random_bits: Callable
+  fold_in: Callable
+
+  def pprint(self):
+    return (pp(self.__class__.__name__) >> pp(':')) + vcat([
+        pp(k) >> pp(' = ') >> pp(v) for k, v in self._asdict().items()
+    ]).indent(2)
+
+
+# -- PRNG key arrays --
+
+def _is_prng_key_data(impl, keys: jnp.ndarray) -> bool:
+  ndim = len(impl.key_shape)
+  try:
+    return (keys.ndim >= 1 and
+            keys.shape[-ndim:] == impl.key_shape and
+            keys.dtype == np.uint32)
+  except AttributeError:
+    return False
+
+@tree_util.register_pytree_node_class
+class PRNGKeyArray:
+  """An array whose elements are PRNG keys.
+
+  This class lifts the definition of a PRNG, provided in the form of a
+  ``PRNGImpl``, into an array-like pytree class. Instances of this
+  class behave like an array whose base elements are keys, hiding the
+  fact that keys are typically arrays (of ``uint32`` dtype) themselves.
+
+  PRNGKeyArrays are also restricted relative to JAX arrays in that
+  they do not expose arithmetic operations. They instead expose
+  wrapper methods around the PRNG implementation functions (``split``,
+  ``random_bits``, ``fold_in``).
+  """
+
+  impl: PRNGImpl
+  keys: jnp.ndarray
+
+  def __init__(self, impl, key_data: jnp.ndarray):
+    # key_data might be a placeholder python `object` or `bool`
+    # instead of a jnp.ndarray due to tree_unflatten
+    if (type(key_data) not in [object, bool] and
+        not _is_prng_key_data(impl, key_data)):
+      raise TypeError(
+          f'Invalid PRNG key data {key_data} for PRNG implementation {impl}')
+    self.impl = impl
+    self.keys = key_data
+
+  def tree_flatten(self):
+    return (self.keys,), self.impl
+
+  @classmethod
+  def tree_unflatten(cls, impl, keys):
+    keys, = keys
+    return cls(impl, keys)
+
+  @property
+  def dtype(self):
+    # TODO(frostig): remove after deprecation window
+    if config.jax_enable_custom_prng:
+      raise AttributeError("'PRNGKeyArray' has no attribute 'dtype'")
+    else:
+      warnings.warn(
+          'deprecated `dtype` attribute of PRNG key arrays', FutureWarning)
+      return np.uint32
+
+  def shape(self):
+    # TODO(frostig): simplify once we always enable_custom_prng
+    if config.jax_enable_custom_prng:
+      return self._shape
+    else:
+      warnings.warn(
+          'deprecated `shape` attribute of PRNG key arrays. In a future version '
+          'of JAX this attribute will be removed or its value may change.',
+          FutureWarning)
+      return self.keys.shape
+
+  @property
+  def _shape(self):
+    base_ndim = len(self.impl.key_shape)
+    return self.keys.shape[:-base_ndim]
```

Comment:
Should we assert that `self._shape` is always equal to `self.impl.key_shape`? Do we ever expect them to be different?

Reply:
The two are different: `_shape` should be the shape of the encompassing "array of keys," whereas `impl.key_shape` is the shape of an individual key. The underlying buffer holding all the key data is an array of shape `(*self._shape, *self.impl.key_shape)`.